### PR TITLE
Onboard mTLS support with AIP-4114 for gsutil

### PR DESCRIPTION
**Pull request by shinfan**
- The certificate discovery has the following logic:
   - Use the user specified command in `.boto` file if available
   - Otherwise, automatically locate the SecureConnectHelper Metadata JSON

Notes: This PR is backwards-compatible

For more information about the mTLS spec please read: [AIP-4114](https://google.aip.dev/auth/4114)

### CODE REVIEW

File: gslib/context_config.py
Comment (NickGoog) [inline gslib/context_config.py:153]: Please add a period at the end of the sentence. Google Python style guide likes perfect grammar.

File: gslib/context_config.py
Comment (NickGoog) [inline gslib/context_config.py:160]: Ditto

File: gslib/context_config.py
Comment (NickGoog) [inline gslib/context_config.py:142]: Docstring needs periods at the end of this line, the "str:" returns line, and the "CertProvisionError" line.

File: gslib/context_config.py
Comment (NickGoog) [inline gslib/context_config.py:34]: Add a period at the end of the comment please

File: gslib/context_config.py
Comment (NickGoog) [inline gslib/context_config.py:119]: Hey, I shouldn't post internal links here, so can you search and check the internal Python style guide for docstring format?

It should be something like:

```
def _check_path():
  """Checks for content aware metadata.

  If content aware metadata exists, return its absolute path;
  otherwise, returns None.

  Returns:
    str: Absolute path if exists. Otherwise, None.
```

File: gslib/context_config.py
Comment (NickGoog) [inline gslib/context_config.py:130]: Needs line breaks above "Returns" and "Raises"

File: gslib/context_config.py
Comment (NickGoog) [inline gslib/context_config.py:145]: Same docstring formatting issue

File: gslib/context_config.py
Comment (NickGoog) [inline gslib/context_config.py:187]: Does this need to be uncommented?

File: gslib/context_config.py
Comment (NickGoog) [inline gslib/context_config.py:35]: Maybe add "DEFAULT" to this constant name?

File: gslib/context_config.py
Comment (NickGoog) [inline gslib/context_config.py:137]: Why not `return json.load(f)`?

File: gslib/context_config.py
Comment (NickGoog) [inline gslib/context_config.py:151]: I would restructure this as:

```
if not metadata_path:
  raise

metadata_json = _read...
```

That way, you can unnest most of the code, and you can stop reading earlier if you know _check_path returned false. Speaking of which, I would recommend differentiating the two `CertProvisionError` messages a little.

File: gslib/context_config.py
Comment (NickGoog) [inline gslib/context_config.py:159]: Wouldn't this be appending to a string, based on http://go/aip/auth/4114 ?

You can only append to lists in Python. Strings use `+` for concatenation.

File: gslib/context_config.py
Comment (NickGoog) [inline gslib/context_config.py:218]: Why indent this?

**Comment by NickGoog** [comment]
This will probably require unit test changes in https://github.com/GoogleCloudPlatform/gsutil/blob/master/gslib/tests/test_context_config.py

But we can hold off on implementing those until the main body of logic is finalized.

File: gslib/context_config.py
Comment (shinfan) [reply gslib/context_config.py:159]: The '--with-passphrase' in this case for security purposes.

https://github.com/googleapis/google-auth-library-python/blob/e6976879b392508c022610ab3ea2ea55c7089c63/google/auth/transport/_mtls_helper.py#L176

**Comment by shinfan** [comment]
> This will require unit test changes in https://github.com/GoogleCloudPlatform/gsutil/blob/master/gslib/tests/test_context_config.py
> 
> But we can hold off on implementing those until the main body of logic is finalized.

I have updated the tests as well.

File: gslib/context_config.py
Comment (NickGoog) [inline gslib/context_config.py:159]: I would write this as "Client certificate provider file not found."

Then, you can write the next `CertProvisionError` as "Client certificate provider  command not found."

File: gslib/context_config.py
Comment (NickGoog) [reply gslib/context_config.py]: Please explain this better.

First, I don't understand how you're using `append` between two strings. That should error:
```
Python 3.9.2 (default, Feb 28 2021, 17:03:44) 
[GCC 10.2.1 20210110] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> 'abc'.append('123')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'str' object has no attribute 'append'
```

Second, based on this line of the auth/4114, it looks like `cert_provider_command` is any user inputted command:
```
"cert_provider_command": "[absolute_path_to_provider_command] --fetch_client_cert"
```

How do we know the command will support "--with-passphrase"? The proposal doesn't require it: http://go/aip/auth/4114

File: gslib/tests/test_context_config.py
Comment (NickGoog) [inline gslib/tests/test_context_config.py:34]: Should this be named something like `DEFAULT_CERT_PROVIDER_FILE_CONTENTS`?

File: gslib/tests/test_context_config.py
Comment (NickGoog) [inline gslib/tests/test_context_config.py:233]: Why were these three tests deleted?

File: gslib/tests/test_context_config.py
Comment (NickGoog) [inline gslib/tests/test_context_config.py:203]: The test naming format is "testSomethingDoesSomething".
For example, "testCustomCommandGeneratesCorrectCert".

**Comment by NickGoog** [comment]
Please add tests that trigger the different `CertProvisionError`s and a test for the "space in path to command" issue you mentioned (e.g. "~/folder with spaces/cert_provider_command.sh").

File: gslib/tests/test_context_config.py
Comment (shinfan) [reply gslib/tests/test_context_config.py:233]: The password flag is no longer related to the custom command scenario so these tests are no longer needed.

File: gslib/context_config.py
Comment (shinfan) [reply gslib/context_config.py]: 1. The field 'cert_provider_command' is a list of strings, so I think `json.load(..)` returns a list in this case.
2. The metadata file and its associated command is Google provided so this '--with-passphrase' assumption should be safe to make.

Comment (shinfan) [reply gslib/context_config.py]: I can make a change to the AIP to indicate that `--with-passphrase` should be supported by the command.

Comment (NickGoog) [reply gslib/context_config.py]: 1) Maybe the AIP should also be updated to show a list of strings for "cert_provider_command"? It looks like a single string right now under "Obtaining the Default Device Certificate". The flag "--fetch_client_cert" is also unexplained.

2) Gotcha.

File: gslib/tests/test_context_config.py
Comment (NickGoog) [inline gslib/tests/test_context_config.py:242]: "testExecutesCustomProviderCommandInBotoConfig"

File: gslib/tests/test_context_config.py
Comment (NickGoog) [inline gslib/tests/test_context_config.py:196]: "testExecutesDefaultProviderCommandFromFile"

This ordering of words seems more accurate. Can you adjust some of the below test names as well?

File: gslib/tests/test_context_config.py
Comment (NickGoog) [inline gslib/tests/test_context_config.py:228]: "testDefaultProviderCommandRaisesNoCommandError"

File: gslib/tests/test_context_config.py
Comment (NickGoog) [inline gslib/tests/test_context_config.py:212]: "testExecutesProviderCommandWithSpaceFromDefaultFile"

**Comment by NickGoog** [comment]
Please write tests confirming the three `CertProvisionError`s are raised at the right time

**Comment by thomasmaclean** [comment]
I just want to note here, that this will feature have to wait until the following week's (July 27th) Cloud SDK release, as there isn't enough time this week to do the final review and run through the release process before the Cloud SDK cutoff on Friday 9am EST.

**Comment by thomasmaclean** [comment]
@shinfan Any updates? I'd like to include this in this coming week's release.

**Comment by shinfan** [comment]
@thomasmaclean Thanks. PTAL.

**Comment by shinfan** [comment]
@thomasmaclean I am not sure why the kokoro is failing. From the log at https://fusion2.corp.google.com/invocations/731f1f30-ba29-4f68-942d-e66eb6a13896/targets/cloud_storage_gsutil%2Flinux%2Fpy35_json/log

The error is ERROR 0719 14:36:16.872922 context_config.py] Failed to provision client certificate: [Errno 2] No such file or directory: '/bin/cat /tmpfs/src/keystore/74008_mtls_test_cert'

Is it the test cert file not uploaded correctly?

File: gslib/context_config.py
Comment (dilipped) [inline gslib/context_config.py:211]: The kokoro failures are happening because we are no longer splitting the command here. Any reason why this was removed?

Comment (dilipped) [reply gslib/context_config.py:211]: Oh, I see that you cert command can be a list if it is not defined in the boto config file as per your change. But if it is defined in the config file, it will be a list. 
So you will have t omake sure that cert_command is of the same type, either a string or a list.

after line 205, you can add the split call.


For kokoro test, we are passing a cat <filename> command here https://github.com/GoogleCloudPlatform/gsutil/blob/5317bc2f19c2c6be8e0a502f209f369c39eb9d2f/test/ci/kokoro/config_generator.sh#L49

**Review by dilipped — CHANGES_REQUESTED** [review]

File: gslib/context_config.py
Comment (shinfan) [reply gslib/context_config.py:214]: You are right. I put the split(' ') back for the boto config command. Thanks.

**Review by dilipped — APPROVED** [review]

**Comment by dilipped** [comment]
I only reviewed the specific change that was making kokoro to fail because of cert command string parsing. Will defer to @NickGoog for final approval as he had left a bunch of comments here.

**Comment by NickGoog** [comment]
Thanks @dilipped,

Waiting on three unit tests to confirm `CertProvisionError` is raised in the three correct places.
Also, didn't we decide to remove `split()`, so that commands with a space in the path would work? (e.g. `/pa th/command.sh`)

**Comment by thomasmaclean** [comment]
@shinfan There was a fairly straightforward merge conflict preventing me from merging in this change which I manually patched, and now there are test failures coming back from that file. I attempted to put the file back to the re-merged state, but you may just want to `git reset 4f33b68fb1037edf9e18183c012805c3ada15a2f --hard` to get back to the pre-merged state and try again yourself. Sorry for any confusion here!

**Review by NickGoog — APPROVED** [review]

**Pull request merged**
